---
title: "1.Processing"
author: "Jack Gordon"
date: "2024-04-15"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "/Users/jackgordon/Documents/DPhil/RNA-Sequencing/Patch-seq")
```


Creates single cell experiment object from kallisto outputs.\
Removes low-qualilty cells.\
Normalises counts with scran.\
Plots qc metrics.\

###1. Processing

```{r, Processing, message = FALSE, warning= FALSE}
#Load libraries
library(tximport)
library(readxl)
library(tidyr)
library(dplyr)
library(rhdf5)
library(SingleCellExperiment)
library(scuttle)
library(AnnotationHub)
library(scater)
library(scran)
library(limma)
library(plotrix)

#Organise directory containing files
dir <- file.path("kallisto")
samples <- read_xlsx(file.path(dir, "Samples.xlsx"))
samples$`Progenitor Origin`[samples$Marker == 'GFP'] <- 'aIP-derived'
samples$`Progenitor Origin`[samples$Marker == 'RFP'] <- 'OP-derived'
files <- file.path(dir, samples$File, "abundance.h5")
names(files) <- samples$SampleID
tx2gene <- read.table(file.path(dir, 'transcripts_to_genes.txt')) #From https://github.com/pachterlab/kallisto-transcriptome-indices/releases

#Get cell:count matrix
txi.kallisto <- tximport(files, type = "kallisto", tx2gene = tx2gene, txOut = FALSE)
head(txi.kallisto$counts)

#Create sinlge cell experiment object
sce <- SingleCellExperiment(assays = list(counts = txi.kallisto$counts))

#Read in metadata for cells
coldata <- DataFrame(samples[, 'SampleID'],
                     samples[,'Mouse'],
                     samples[,'Age'],
                     samples[,'Marker'], 
                     samples[, 'Progenitor Origin'])

sce <- SingleCellExperiment(assays = list(counts=txi.kallisto$counts), colData=coldata)
unprocessed <- sce
saveRDS(unprocessed, file = "unprocessed_sce.rds")
sce <- readRDS("data/unprocessed_sce.rds")

#Convert ensemblIDs to Gene symbols
ens.mm.v97 <- AnnotationHub()[["AH73905"]] #Load in reference genome
rownames(sce) <- gsub("\\..*","", rownames(sce)) #Remove version number (after .) in ENSEMBL IDs

rowData(sce)$ENSEMBL <- rownames(sce) #Store ensembl IDs in rowdata

rowData(sce)$SYMBOL <- mapIds(ens.mm.v97, keys=rownames(sce),
                                   keytype="GENEID", column="SYMBOL") #Store gene symbols in rowdata

rowData(sce)$SEQNAME <- mapIds(ens.mm.v97, keys=rownames(sce),
                                    keytype="GENEID", column="SEQNAME") #Store seqname in rowdata

rownames(sce) <- uniquifyFeatureNames(rowData(sce)$ENSEMBL, 
                                           rowData(sce)$SYMBOL) #Use symbol as rowname unless only ensembl available

#Convert NAs to "Unlabelled"
sce$Progenitor.Origin <- sce$Progenitor.Origin %>% replace_na("Unlabelled") 

#Convert marker to factor
sce$Marker <- factor(sce$Marker)

#Quality Control

#Find mitochondrial genes
mito <- which(rowData(sce)$SEQNAME=="MT")

#Calcuate QC metrics per cell, incl. mt percent
stats <- perCellQCMetrics(sce, subsets=list(Mt=mito))
qc <- quickPerCellQC(stats, percent_subsets=c("subsets_Mt_percent"), batch=sce$Marker)

#Add metrics to sce object
sce <- addPerCellQCMetrics(sce, subsets=list(Mt=mito))

#Bind stats to sce object colda
colData(sce) <- cbind(colData(sce), stats)


#Add discard variable based on number og genes
sce$discard <- qc$low_n_features 

#Save unfiltered sce with qc metrics
unfiltered <- sce 
saveRDS(unfiltered, file = "data/unfiltered_sce.rds") #Save unfiltered dataset after adding QC info


#Remove low quality cells and save sce object
filtered <- sce[,!qc$low_n_features]
saveRDS(filtered, file = "data/filtered_sce.rds")

#Set sce as filtered
sce <- filtered #Save filtered dataset after removing low quality cells
```

Plot qc metrics

```{r qc plots, message = FALSE, warning= FALSE}
#Load libraries
library(ggpubr)
library(ggplot2)
library(scales)

#Get coldata dataframe with qc metrics
filtered <- as.data.frame(colData(filtered))

#Set plotting aesthetics
#Set colors and levels for progenitor.origin variables and comparisons for stat_summary
levels(filtered$Progenitor.Origin) <- c("aIP-derived", "OP-derived", "Unlabelled")
level_order <- c("OP-derived", "aIP-derived", "Unlabelled")
cols <- c('forestgreen', 'firebrick3', 'darkgray')
names(cols) = levels(filtered$Progenitor.Origin)

#Plot number of reads
ggbarplot(filtered, x = "Progenitor.Origin", y = "sum", fill = "Progenitor.Origin", 
               palette = cols, add = "mean_se", order = c("OP-derived", "aIP-derived", "Unlabelled"),
               error.plot = "upper_errorbar", add.params = list(color = "black", alpha = 0.5)) +
  theme_classic(base_size = 15) + theme(legend.position = "none") + xlab("") + ylab("Mapped Reads (millions)") +
  scale_y_continuous(labels = unit_format(unit = "M", scale = 1e-6))

#Plot detected genes
ggbarplot(filtered, x = "Progenitor.Origin", y = "detected", fill = "Progenitor.Origin", 
               palette = cols, add = "mean_se", order = c("OP-derived", "aIP-derived", "Unlabelled"),
               error.plot = "upper_errorbar", add.params = list(color = "black", alpha = 0.5)) +
  theme_classic(base_size = 15) + theme(legend.position = "none") + xlab("") + ylab("Detected Genes (thousands)") +
  scale_y_continuous(labels = unit_format(unit = "K", scale = 1e-3)) 

#Plot percent.mt
ggbarplot(filtered, x = "Progenitor.Origin", y = "subsets_Mt_percent", fill = "Progenitor.Origin", 
               palette = cols, add = "mean_se", order = c("OP-derived", "aIP-derived", "Unlabelled"),
               error.plot = "upper_errorbar", add.params = list(color = "black", alpha = 0.5)) +
  theme_classic(base_size = 15) + theme(legend.position = "none") + xlab("") + ylab("Mitochondrial Reads (%)") +
  scale_y_continuous(breaks = c(0, 2, 4, 6, 8, 10)) 

```

```{r, sessionInfo}
sessionInfo()
```

